Extensions/XMPPMUCSUB/XMPPMUCSUB.m

```diff
@@ -0,0 +1,789 @@
+    }
+    return self;
+}
+
+- (BOOL)activate:(XMPPStream *)aXmppStream
+{
+    if ([super activate:aXmppStream])
+    {
+        XMPPLogVerbose(@"%@: Activated", THIS_FILE);
+        
+        xmppIDTracker = [[XMPPIDTracker alloc] initWithStream:xmppStream
+                                                dispatchQueue:moduleQueue];
+        
+#ifdef _XMPP_CAPABILITIES_H
+        [xmppStream autoAddDelegate:self
+                      delegateQueue:moduleQueue
+                   toModulesOfClass:[XMPPCapabilities class]];
+#endif
+        return YES;
+    }
+    
+    return NO;
+}
+
+- (void)deactivate
+{
+    XMPPLogTrace();
+    
+    dispatch_block_t block = ^{ @autoreleasepool {
+        [self->xmppIDTracker removeAllIDs];
+        self->xmppIDTracker = nil;
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag))
+        block();
+    else
+        dispatch_sync(moduleQueue, block);
+    
+#ifdef _XMPP_CAPABILITIES_H
+    [xmppStream removeAutoDelegate:self delegateQueue:moduleQueue fromModulesOfClass:[XMPPCapabilities class]];
+#endif
+    
+    [super deactivate];
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark Public API
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+/**
+ * This method provides functionality of Discovering a MUCSUB Service.
+ *
+ * @link {https://docs.ejabberd.im/developer/xmpp-clients-bots/extensions/muc-sub/#discovering-support}
+ *
+ * Example 1. Entity Queries Server for Associated Services
+ *
+ * <iq from='hag66@shakespeare.example/pda'
+ *       to='muc.shakespeare.example'
+ *       type='get'
+ *       id='ik3vs715'>
+ *   <query xmlns='http://jabber.org/protocol/disco#info'/>
+ * </iq>
+ */
+- (void)discoverFeatures
+{
+    // This is a public method, so it may be invoked on any thread/queue.
+    
+    dispatch_block_t block = ^{ @autoreleasepool {
+        if (self->hasRequestedFeatures) return; // We've already requested services
+        
+        NSString *mucService = [NSString stringWithFormat:@"conference.%@", self->xmppStream.myJID.domain];
+        NSXMLElement *query = [NSXMLElement elementWithName:@"query"
+                                                      xmlns:XMPPMucSubDiscoInfo];
+        XMPPIQ *iq = [XMPPIQ iqWithType:@"get"
+                                     to:[XMPPJID jidWithString:mucService]
+                              elementID:[self->xmppStream generateUUID]
+                                  child:query];
+        
+        [self->xmppIDTracker addElement:iq
+                                 target:self
+                               selector:@selector(handleDiscoverFeaturesQueryIQ:withInfo:)
+                                timeout:60];
+        
+        [self->xmppStream sendElement:iq];
+        self->hasRequestedFeatures = YES;
```

Comment:
Should this be reset outside of the block? Can only run this 1 time the entire lifecycle of the app?

Reply:
Oh I see it now... haha.

Reply:
No it can be run many times. This will prevent sending another request while we have sent one and are waiting for the response

Reply:
I guess it can be outside of block but I like the fact that it set the flag only after we send iq.